### PR TITLE
feat(billing): add usage endpoint

### PR DESCRIPTION
**Pull request by domhhv**
Updates #1348 

## Summary by CodeRabbit

- **New Features**
  - Introduced API endpoints to retrieve detailed usage history and aggregated usage statistics for wallet addresses, including daily and cumulative spending in AKT, USDC, and USD.
  - Added input validation for addresses and date ranges, with clear error responses for invalid queries.
  - Added a new method to count distinct leases by owner for enhanced usage statistics.

- **Bug Fixes**
  - Improved address generation in test data to ensure valid Bech32-encoded addresses.

- **Tests**
  - Added comprehensive unit and functional tests for usage history and statistics endpoints, including edge cases and error handling.
  - Introduced new data seeders to support robust test scenarios.

### CODE REVIEW

File: apps/api/src/billing/repositories/usage/usage.repository.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+interface BillingUsageRawResult {
+  Date: string;
+  "Active Leases": number;
+  "Daily AKT Spent": number;
+  "Total AKT Spent": number;
+  "Daily USDC Spent": number;
+  "Total USDC Spent": number;
+  "Daily USD Spent": number;
+  "Total USD Spent": number;
```
Comment (ygrishajev) [inline apps/api/src/billing/repositories/usage/usage.repository.ts]: Non-blocking: this notation looks weird as internal query. I'd suggest using smth more usual like snake/camel case.

File: apps/api/src/billing/services/usage/usage.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+import { UsageService } from "./usage.service";
+
+import { AkashAddressSeeder } from "@test/seeders/akash-address.seeder";
+
+describe(UsageService.name, () => {
+  let service: UsageService;
+  let usageRepository: jest.Mocked<UsageRepository>;
+  let leaseRepository: jest.Mocked<LeaseRepository>;
+
+  beforeEach(() => {
```
Comment (ygrishajev) [inline apps/api/src/billing/services/usage/usage.service.spec.ts]: Pls use setup function approach, avoid jest hooks. This is smth we follow for all cases.

File: apps/api/src/billing/services/usage/usage.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+    service = new UsageService(usageRepository, leaseRepository);
+  });
+
+  describe("getHistory", () => {
+    const address = AkashAddressSeeder.create();
+    const startDate = "2024-01-01";
+    const endDate = "2024-01-31";
+
+    const mockUsageData = [
```
Comment (ygrishajev) [inline apps/api/src/billing/services/usage/usage.service.spec.ts]: This looks like a case for some seeder with faker.

File: apps/api/src/billing/services/usage/usage.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        dailyAktSpent: 2.1,
+        totalAktSpent: 10.8,
+        dailyUsdcSpent: 1.0,
+        totalUsdcSpent: 4.75,
+        dailyUsdSpent: 3.1,
+        totalUsdSpent: 15.55
+      })
+    ];
+
+    const usageRepository = {
```
Comment (ygrishajev) [inline apps/api/src/billing/services/usage/usage.service.spec.ts]: There are utilities to avoid this type casting and easy mocking

https://github.com/akash-network/console/blob/main/apps/api/test/services/stub.ts

jest-mock-extended https://github.com/akash-network/console/blob/main/apps/api/src/healthz/services/healthz/healthz.service.spec.ts

Comment (domhhv) [reply apps/api/src/billing/services/usage/usage.service.spec.ts]: Thanks, used `stub` utility

File: apps/api/src/billing/services/usage/usage.service.spec.ts (hunk 3 above)
Comment (domhhv) [reply apps/api/src/billing/services/usage/usage.service.spec.ts]: Created and used `BillingUsageSeeder` in such cases

File: apps/api/src/billing/services/usage/usage.service.spec.ts (hunk 2 above)
Comment (domhhv) [reply apps/api/src/billing/services/usage/usage.service.spec.ts]: Done

File: apps/api/src/billing/repositories/usage/usage.repository.ts (hunk 1 above)
Comment (domhhv) [reply apps/api/src/billing/repositories/usage/usage.repository.ts]: +1
Used camelCase

File: apps/api/src/billing/controllers/usage/usage.controller.spec.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      averageLeasesPerDay: 1.5
+    };
+
+    service.getHistoryStats.mockResolvedValue(output);
+
+    const result = await controller.getHistoryStats(address, startDate, endDate);
+
+    expect(service.getHistoryStats).toHaveBeenCalledWith(address, startDate, endDate);
+    expect(result).toEqual(output);
+  });
```
Comment (stalniy) [inline apps/api/src/billing/controllers/usage/usage.controller.spec.ts]: thought(non-blocking): based on the implementation, I see that controller just forwards the call to underlying service. So, there is no logic to test and I would not add UNIT tests like this at all because they give very little value.

Each time we refactor controller, we also will need to refactor this test. Even in case of minor changes, like adding additional field to returned result.

So, I'd recommend not to do this in future. Controllers are the best covered with functional tests, then we have higher confidence that this thing work. And it will make our lives easier, if some day we will decide to migrate to Nestjs

File: apps/api/src/billing/http-schemas/usage.schema.ts [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    startDate: z.string().date().optional().openapi({
+      description: "Start date (YYYY-MM-DD). Defaults to 30 days before endDate",
+      example: "2024-01-01"
+    }),
+    endDate: z
+      .string()
+      .date()
+      .default(() => new Date().toISOString().split("T")[0])
+      .openapi({
+        description: "End date (YYYY-MM-DD). Defaults to today",
```
Comment (stalniy) [inline apps/api/src/billing/http-schemas/usage.schema.ts]: **suggestion(non-blocking):** I'd note that "today" is by UTC 23:59:59

File: apps/api/src/billing/http-schemas/usage.schema.ts [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    }
+  )
+  .refine(
+    data => {
+      return isValidBech32Address(data.address, "akash");
+    },
+    {
+      message: "Invalid Akash address format"
+    }
+  );
```
Comment (stalniy) [inline apps/api/src/billing/http-schemas/usage.schema.ts]: question(non-blocking): is this possible to connect this with corresponding field?
 
Like this:
```ts
address: z.string().min(1).openapi({
      description: "The wallet address to get billing and usage data for",
      example: "akash18andxgtd6r08zzfpcdqg9pdr6smks7gv76tyt6"
    }).refine(
    data => {
      return isValidBech32Address(data.address, "akash");
    },
    {
      message: "Invalid Akash address format"
    }
  )
```

File: apps/api/src/billing/services/usage/usage.service.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    private readonly leaseRepository: LeaseRepository
+  ) {}
+
+  async getHistory(address: string, startDate: string, endDate: string): Promise<UsageHistoryResponse> {
+    return await this.usageRepository.getHistory(address, startDate, endDate);
+  }
+
+  async getHistoryStats(address: string, startDate: string, endDate: string): Promise<UsageHistoryStats> {
+    const historyData = await this.usageRepository.getHistory(address, startDate, endDate);
+    const totalLeases = await this.leaseRepository.countByOwner(address);
```
Comment (stalniy) [inline apps/api/src/billing/services/usage/usage.service.ts]: suggestion(non-blocking): it seems this can be done in parallel

File: apps/api/test/seeders/akash-address.seeder.ts [hunk 9]
```diff
@@ -1,7 +1,14 @@
...
 export class AkashAddressSeeder {
   static create(): string {
-    return `akash${faker.string.alphanumeric({ length: 39 })}`;
+    const addressData = new Uint8Array(20);
+
+    for (let i = 0; i < 20; i++) {
+      addressData[i] = faker.number.int({ min: 0, max: 255 });
+    }
+
+    return toBech32("akash", addressData);
```
Comment (stalniy) [inline apps/api/test/seeders/akash-address.seeder.ts:12]: **praise:** 👍 thanks

File: apps/api/src/billing/services/usage/usage.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+
+        expect(result.totalSpent).toBe(427.22);
+        expect(result.averagePerDay).toBe(213.61);
+        expect(result.totalLeases).toBe(1000);
+        expect(result.averageLeasesPerDay).toBe(500.0);
+      });
+    });
+  });
+
+  function setup(usageData?: BillingUsageRawResult[], totalLeases?: number) {
```
Comment (stalniy) [inline apps/api/src/billing/services/usage/usage.service.spec.ts]: suggestion(non-blocking): better to have single `input` parameter and inline its type:

```ts
function setup(input?: {
  usageData?: BillingUsageRawResult[], 
  totalLeases?: number
}) {}
```

File: apps/api/src/billing/services/usage/usage.service.spec.ts [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+    });
+
+    const leaseRepository = stub<LeaseRepository>({
+      countByOwner: jest.fn()
+    });
+
+    const service = new UsageService(usageRepository, leaseRepository);
+
+    usageRepository.getHistory.mockResolvedValue(mockUsageData);
+    leaseRepository.countByOwner.mockResolvedValue(leasesCount);
```
Comment (stalniy) [inline apps/api/src/billing/services/usage/usage.service.spec.ts]: suggestion(non-blocking): Please inline this in `stub` creation:

```ts
    const usageRepository = stub<UsageRepository>({
      getHistory: jest.fn().moceResolvedValue(mockUsageData)
    });
```

File: apps/api/src/billing/services/usage/usage.service.spec.ts [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+        dailyAktSpent: 2.1,
+        totalAktSpent: 10.8,
+        dailyUsdcSpent: 1.0,
+        totalUsdcSpent: 4.75,
+        dailyUsdSpent: 3.1,
+        totalUsdSpent: 15.55
+      })
+    ];
+
+    const usageRepository = stub<UsageRepository>({
```
Comment (stalniy) [inline apps/api/src/billing/services/usage/usage.service.spec.ts]: suggestion(non-blocking): I think `stub` is an early helper which was added before `jest-mock-extended`. As we now have jest-mock-extended, we can use its `mock` helper

File: apps/api/test/seeders/day.seeder.ts [hunk 13]
```diff
@@ -17,6 +17,11 @@ export class DaySeeder {
 
   static async createInDatabase(overrides: Partial<CreationAttributes<Day>> = {}): Promise<Day> {
     const seed = DaySeeder.create(overrides);
-    return await Day.create(seed);
+    try {
+      return await Day.create(seed);
+    } catch (error) {
+      console.error("Error creating Day in database:", error);
```
Comment (stalniy) [inline apps/api/test/seeders/day.seeder.ts]: question(non-blocking): why do you need this?

File: apps/api/src/billing/repositories/usage/usage.repository.ts [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+export class UsageRepository {
+  async getHistory(address: string, startDate: string, endDate: string) {
+    const query = `
+      WITH date_range AS (
+        SELECT generate_series(
+          :startDate::date,
+          :endDate::date,
+          '1 day'::interval
+        )::date as date
+      ),
```
Comment (stalniy) [inline apps/api/src/billing/repositories/usage/usage.repository.ts:27]: issue(blocking): this SQL query contains the main logic responsible to calculating and gathering statistic, but there are no tests that covers it.

It's important to ensure this query properly work, otherwise it will be hard to change it

**Review by stalniy — CHANGES_REQUESTED** [review]

File: apps/api/src/billing/http-schemas/usage.schema.ts (hunk 7 above)
Comment (stalniy) [reply apps/api/src/billing/http-schemas/usage.schema.ts]: Actually we have a schema for akash address -> `AkashAddressSchema`: https://github.com/akash-network/console/blob/main/apps/api/src/utils/schema.ts#L5

File: apps/api/src/billing/controllers/usage/usage.controller.spec.ts (hunk 5 above)
Comment (domhhv) [reply apps/api/src/billing/controllers/usage/usage.controller.spec.ts]: I think so too. I removed it for now.

File: apps/api/src/billing/http-schemas/usage.schema.ts (hunk 6 above)
Comment (domhhv) [reply apps/api/src/billing/http-schemas/usage.schema.ts]: Added to description

File: apps/api/src/billing/http-schemas/usage.schema.ts (hunk 7 above)
Comment (domhhv) [reply apps/api/src/billing/http-schemas/usage.schema.ts]: Nice one - reused

File: apps/api/src/billing/services/usage/usage.service.spec.ts (hunk 10 above)
Comment (domhhv) [reply apps/api/src/billing/services/usage/usage.service.spec.ts]: Makes sense, refactored

File: apps/api/src/billing/services/usage/usage.service.spec.ts (hunk 12 above)
Comment (domhhv) [reply apps/api/src/billing/services/usage/usage.service.spec.ts]: Done

File: apps/api/src/billing/services/usage/usage.service.spec.ts (hunk 11 above)
Comment (domhhv) [reply apps/api/src/billing/services/usage/usage.service.spec.ts]: Done

File: apps/api/src/billing/services/usage/usage.service.ts (hunk 8 above)
Comment (domhhv) [reply apps/api/src/billing/services/usage/usage.service.ts]: Right, it surely can - used Promise.all

File: apps/api/test/seeders/akash-address.seeder.ts (hunk 9 above)
Comment (domhhv) [reply apps/api/test/seeders/akash-address.seeder.ts:12]: Thank you!

File: apps/api/test/seeders/day.seeder.ts (hunk 13 above)
Comment (domhhv) [reply apps/api/test/seeders/day.seeder.ts]: Ouch, we don't. This was some debugging I added when my functional tests were failing and then totally forgot to remove. Thanks for the catch!

File: apps/api/src/billing/repositories/usage/usage.repository.ts (hunk 14 above)
Comment (domhhv) [reply apps/api/src/billing/repositories/usage/usage.repository.ts:27]: Sounds good, I've created usage.repository.spec.ts which covers absent and present leases with various denominations. I tried to base it on an existing spec - provider.repository.spec.ts - which appears to be the only spec for a repository and is quite recent. Both specs use a straightforward setup function and an afterEach hook. I'm open to any suggestions if the new one could be clearer or more comprehensive.

File: apps/api/src/billing/repositories/usage/usage.repository.spec.ts [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+    let createdDeploymentIds: string[] = [];
+    let createdDeploymentGroupIds: string[] = [];
+    let createdProviderAddresses: string[] = [];
+
+    afterEach(async () => {
+      await chainDb.models.lease.destroy({
+        where: {
+          id: { [Op.in]: createdLeaseIds }
+        }
+      });
```
Comment (stalniy) [inline apps/api/src/billing/repositories/usage/usage.repository.spec.ts:34]: **suggestion(non-blocking):** maybe for time being, we could create 

```ts
function createEntityTracker<T extends Model>(options: { 
  create: (attrs: CreationAttributes<T>) => Promise<T>, 
  destroy: (keys: unknown[]) => Promise<void> 
}) {
  const primaryKeys = new Set<unknown>()
  return {
    async create<T>(attrs: CreationAttributes<T>): Promise<T> {
       const entity = await options.create(attrs);
       primaryKeys.add(getSomehowPrimaryKey(entity));
       return entity;
    },
    async destroy() {
      await options.destroy(primaryKeys);
    }
  }
}
```

Comment (stalniy) [reply apps/api/src/billing/repositories/usage/usage.repository.spec.ts:34]: ok, never mind. Need to check libraries like fishery and configure long term solution for this

**Review by stalniy — APPROVED** [review]

File: apps/api/src/billing/controllers/usage/usage.controller.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
+import { singleton } from "tsyringe";
+
+import { UsageHistoryResponse, UsageHistoryStats } from "@src/billing/http-schemas/usage.schema";
+import { UsageService } from "@src/billing/services/usage/usage.service";
+
+@singleton()
+export class UsageController {
+  constructor(private readonly usageService: UsageService) {}
+
+  async getHistory(address: string, startDate: string, endDate: string): Promise<UsageHistoryResponse> {
```
Comment (baktun14) [inline apps/api/src/billing/controllers/usage/usage.controller.ts:10]: We should memoize these endpoints to something like few seconds, maybe 15 seconds
